c/tskit/trees.c

```diff
@@ -2076,19 +2076,651 @@ typedef struct {
+    const tsk_id_t *restrict mut_nodes = self->tables->mutations.node;
+    const tsk_site_t *restrict site;
+
+    tsk_id_t u, node, top_mut_node, *stack;
+    tsk_bit_array_t mut_samples, mut_samples_row, node_paths, path, out_row;
+    tsk_size_t num_mutations, mut_samples_offset, mut_offset;
+    int stack_top;
+
+    // Get the number of mutations and the top mutation node
+    num_mutations = 0;
+    top_mut_node = 0;
+    for (tsk_size_t s = 0; s < num_sites; s++) {
+        site = &tree_sites[s];
+        for (tsk_size_t m = 0; m < site->mutations_length; m++) {
+            num_mutations++;
+            if (site->mutations[m].node > top_mut_node) {
+                top_mut_node = site->mutations[m].node;
+            }
+        }
+    }
+
+    tsk_bit_array_init(&mut_samples, num_samples, num_mutations);
+    if (ret != 0) {
+        goto out;
+    }
+    tsk_bit_array_init(&node_paths, tsk_treeseq_get_num_nodes(self), num_mutations);
+    if (ret != 0) {
+        goto out;
+    }
+
+    // Size of stack is derived from tsk_tree_get_size_bound
+    stack = tsk_malloc((1 + self->num_samples + num_edges) * sizeof(*stack));
+    if (stack == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    stack_top = 0;
+    stack[stack_top] = parent[top_mut_node];
+
+    // Current position minus num ancestral alleles to this point, if there are any sites
+    mut_offset = num_sites
+                     ? (site_offsets[tree_sites[0].id] - (tsk_size_t) tree_sites[0].id)
+                     : 0;
+
+    // Traverse the current tree, tracking which samples are under each mutation
+    while (stack_top >= 0) {
+        node = stack[stack_top];
+        stack_top--;
+        for (tsk_size_t m = 0; m < num_mutations; m++) {
+            tsk_bit_array_get_row(&node_paths, m, &path);
+            tsk_bit_array_get_row(&mut_samples, m, &mut_samples_row);
+            if (mut_nodes[m + mut_offset] == node
+                || tsk_bit_array_contains_id(&path, parent[node])) {
+                tsk_bit_array_add_bit(&path, (tsk_bit_array_value_t) node);
+                if (flags[node] & TSK_NODE_IS_SAMPLE) {
+                    tsk_bit_array_add_bit(
+                        &mut_samples_row, (tsk_bit_array_value_t) node);
+                }
+            }
+        }
+        u = right_child[node];
+        while (u != TSK_NULL) {
+            stack_top++;
+            stack[stack_top] = u;
+            u = left_sib[u];
+        }
+    }
+
+    mut_samples_offset = 0;
+    for (tsk_size_t s = 0; s < num_sites; s++) {
+        site = &tree_sites[s];
+
+        tsk_bit_array_get_row(&mut_samples, mut_samples_offset, &mut_samples_row);
+        mut_samples_offset += self->site_mutations_length[site->id];
+
+        tsk_bit_array_get_row(allele_samples, site_offsets[site->id], &out_row);
+        get_allele_samples(site, &mut_samples_row, &out_row, &(*num_alleles)[site->id]);
+    }
+out:
+    tsk_safe_free(mut_samples.data);
```

Comment:
Better to call ``tsk_bit_array_free(&mut_samples)``, so that the implementation can change